src/peep/peep.c

```diff
@@ -7772,13 +7772,404 @@ static void peep_on_enter_or_exit_ride(rct_peep *peep, int rideIndex, int flags)
+		// i.e whether they're willing to pay the asking price
+		if (dx < price) {
+			dx -= price;
+			if (shopItem == 4) {
+				if (RCT2_ADDRESS_CURRENT_RAIN_LEVEL != 0)
+					goto loc_69B221;
+			}
+			///loc_69B1F8:
+			dx = -dx;
+			if (peep->happiness < 128)
+				dx /= 2;
+			/// loc_69B204:
+			if (peep->happiness < 180)
+				dx /= 2;
+			/// loc 69B20D:
+			if (dx > (scenario_rand() & 0x07)) {
+				///loc_69B530:
+				// "I'm not paying that much for x"
+				shopItem += 72;
+				if (shopItem > 104)
+					shopItem += 32;
+				peep_insert_new_thought(peep, shopItem, rideIndex); // Possibly superfluous argument (ah is usually rideIndex)
+				return 0;
+			}
+		}
+		else {
+			dx -= price;
+			if (dx < 8)
+				dx = 8;
+			uint8 dl = dx & 0xFF;
+			/// loc_69B1AC
+			if (dl > (scenario_rand() & 0x07)) {
+				if (!(RCT2_GLOBAL(RCT2_ADDRESS_PARK_FLAGS, uint32) & (1 << 11))) {
+					// "This x is a really good value"
+					int thought_item = shopItem + 38;
+					if (thought_item < 70)
+						thought_item += 34;
+					peep_insert_new_thought(peep, shopItem, rideIndex); // Possibly superfluous, same as above
+				}
+			}
+			/// loc_69B1D4:
+			dl *= 4;
+			peep->happiness_growth_rate = min((peep->happiness_growth_rate + dl), 255);
+			/// loc_69B1E0:
+			peep->happiness = min((peep->happiness + dl), 255);
+		}
+	}
+
+loc_69B221:
+	uint32 edi = rideIndex + 0x0260; // This is used way at the end but is assigned here.
+	if (!RCT2_GLOBAL(0x00F1AEE3, unsigned char)) { // If no voucher
+		if (shopItem >= 32) {
+			// This seems to be identical to the construct above.
+			dx = RCT2_ADDRESS(0x00982246, uint16)[shopItem2 * 4];
+			if (RCT2_ADDRESS_CURRENT_TEMPERATURE >= 21)
+				dx = RCT2_ADDRESS(0x00982248, uint16)[shopItem2 * 4];
+			if (RCT2_ADDRESS_CURRENT_TEMPERATURE <= 11)
+				dx = RCT2_ADDRESS(0x0098224A, uint16)[shopItem2 * 4];
+		}
+		else {
+			dx = RCT2_ADDRESS(0x00982166, uint16)[shopItem2 * 4];
+			if (RCT2_ADDRESS_CURRENT_TEMPERATURE >= 21)
+				dx = RCT2_ADDRESS(0x00982168, uint16)[shopItem2 * 4];
+			if (RCT2_ADDRESS_CURRENT_TEMPERATURE <= 11)
+				dx = RCT2_ADDRESS(0x0098216A, uint16)[shopItem2 * 4];
+		}
+		/// loc_69B293:
+		dx -= price;
+		uint8 bl = 0; // Satisfaction incremental
+		if (dx > 0xFFF8) {
+			bl++;
+			if (dx > 0xFFFD) {
+				bl++;
+				if (dx > 3)
+					bl++;
+			}
+		}
+		/// loc_69B2B0:
+		// FIXME: need to make sure the ride pointer is passed in
+		// to be able to call this correctly.
+		// May require different params to be made available from higher up.
+		ride_update_satisfaction(ride, bl);
+	}
+///loc_69B2B7:
+	// The peep has now decided to buy the item (or, specifically, has not
+	// been dissuaded so far).
+	// FIXME: Is there a call to have the peep buy an item?
+	if (shopItem >= 32)
+		peep->item_extra_flags |= (1 << shopItem2);
+	else
+		peep->item_standard_flags |= (1 << shopItem2);
+
+/// loc_69B2C8:
+	if (shopItem == 20) // T-Shirt
+		peep->tshirt_colour = RCT2_ADDRESS(0x1362AB4, uint8)[rideIndex * 0x260];
```

Comment:
It's off by 0x1BC, so I made this educated guess, is it right? `RCT2_ADDRESS(RCT2_ADDRESS_RIDE_LIST, rct_ride)[rideIndex].track_colour_main[0]`